### DISCUSSION TITLE
Implement Loader for &str keys?

### DISCUSSION THREAD

**Discussion by urkle** [async-graphql/async-graphql#1100] [General]
I am trying to implement DataLoader for an &str key if possible (to maybe reduce cloning strings)
Current is like this.
```rust
#[async_trait]
impl Loader<String> for QuotaLoader {
    type Value = Vec<MyObject>;
    type Error = FieldError;

    async fn load(&self, keys: &[String]) -> Result<HashMap<String, Self::Value>, Self::Error> {
        let list: HashMap<_, _> = keys
            .iter()
            .filter_map(|e| my_func(e).map(|q| (e.clone(), q)).ok())
            .collect();

        Ok(list)
    }
}
```

But I have to call the loader like so

```rust
loader.load_one(mykey.clone()).await?
```

I would like to be able to use &str if possible like this.

```rust
#[async_trait]
impl Loader<&str> for QuotaLoader {
    type Value = Vec<MyObject>;
    type Error = FieldError;

    async fn load(&self, keys: &[&str]) -> Result<HashMap<&str, Self::Value>, Self::Error> {
        let list: HashMap<_, _> = keys
            .iter()
            .filter_map(|e| my_func(e).map(|q| (e, q)).ok())
            .collect();

        Ok(list)
    }
}
```

However, all I get is lifetime does not live long enough errors.

error[E0478]: lifetime bound not satisfied
  --> src/loader.rs:20:6
   |
20 | impl Loader<&str> for MyLoader {
   |      ^^^^^^^^^^^^
   |
note: lifetime parameter instantiated with the lifetime `'impl0` as defined here
  --> src/loader.rs:20:1
   |
20 | impl Loader<&str> for MyLoader {
   | ^^^^
   = note: but lifetime parameter must outlive the static lifetime

What is causing the requirement of the static lifetime?

My intention is to have a map loop that re-uses the keys as the key in the hash and the result of a func call (that takes an AsRef<str>) that returns the Vec.

**Comment by sunli829** [comment]
No, cache needs to clone this key, it must be `'static` 🙂
